### ISSUE TITLE
Closures perform better resolution of types than functions

### ISSUE THREAD

**Issue by uzytkownik** [rust-lang/rust#44711]
I would consider those functions to be equivalent:

```rust
with_p4branch(branch, |s| {String::from(s)});
with_p4branch(branch, String::from);
```

However as closure succeeds the second fails with:
```
error[E0281]: type mismatch: `fn(_) -> std::string::String {<std::string::String as std::convert::From<_>>::from}` implements the trait `std::ops::FnOnce<(_,)>`, but the trait `for<'r> std::ops::FnOnce<(&'r str,)>` is required
   --> src/main.rs:106:5
    |
106 |     with_p4branch(branch, String::from)
    |     ^^^^^^^^^^^^^ expected concrete lifetime, found bound lifetime parameter
    |
    = note: required by `with_p4branch`

error[E0271]: type mismatch resolving `for<'r> <fn(_) -> std::string::String {<std::string::String as std::convert::From<_>>::from} as std::ops::FnOnce<(&'r str,)>>::Output == _`
   --> src/main.rs:106:5
    |
106 |     with_p4branch(branch, String::from)
    |     ^^^^^^^^^^^^^ expected bound lifetime parameter, found concrete lifetime
    |
    = note: concrete lifetime that was found is lifetime '_#7r
    = note: required by `with_p4branch`

error: aborting due to previous error(s)
```

I think it should be trivially unified as with closure. I use 1.19.0.

**Comment by bluss** [comment]
Could you please complete this with the exact signature of the `with_p4branch` function? That will help answering further questions.

I see one difference between the two already -- where we see `s` in `String::from(s)` there is a coercion site, so the two pieces of code will be different in that way, and the coercion will help accepting more types in the first piece.

**Comment by uzytkownik** [comment]
```rust
fn with_p4branch<R, F: FnOnce(&str) -> R>(branch: &git2::Branch, f: F) -> Result<R, SomeErrorIDontRemember>;

// Here's how it was used:
fn get_p4branch(branch: &git2::Branch) -> Result<String, SomeErrorIDontRemember> {
    with_p4branch(branch, String::from)
}
```

**Comment by bluss** [comment]
Thanks!

This looks like issue #36582, which means that it's a bug that has already been reported and acknowledged, and will hopefully be fixed at some point. Since the issue looks like it matches exactly, this one is closed as a duplicate of #36582.
